tarantool/crud.py

```diff
@@ -0,0 +1,84 @@
+"""
+This module provides API for interaction with the `crud`_ module.
+
+.. _crud: https://github.com/tarantool/crud/
+"""
+
+from tarantool.error import DatabaseError, ER_NO_SUCH_PROC
+
+
+class CrudResponse(object):
+    """
+    Contains response fields from the `crud`_ module that correspond 
+    to the Lua implementation.
+
+    .. _crud: https://github.com/tarantool/crud/
+    """
+
+    def __init__(self, response):
+        """
+        Sets response fields as in Lua implementation.
+
+        :param response: The response object of the crud module call.
+        :type response: :class:`~tarantool.response.Response`
+        """
+
+        if isinstance(response, dict):
+            for response_field_name in response.keys():
+                if isinstance(response_field_name, bytes):
+                    setattr(self, response_field_name.decode(), response[response_field_name])
+                else:
+                    setattr(self, response_field_name, response[response_field_name])
+        else:
+            raise RuntimeError('Unable to decode response to object due to unknown type')
+
+
+class CrudResult(CrudResponse):
+    """
+    Contains result's fields from result variable 
+    of crud module operation.
+    """
+
+
+class CrudError(CrudResponse):
+    """
+    Contains error's fields from error variable 
+    of crud module operation.
+    """
+
+
+def wrapCrudUndefinedProcedureError(e: DatabaseError):
+    """
+    Supplements information about an unsuccessful attempt 
+    to find the crud module procedure.
+
+    :param e: The error object of the crud module call.
+    :type e: :class:`~tarantool.error.DatabaseError`
+    """
+
+    info = ". Ensure that you're calling crud.router and user has sufficient grants"
+    e.message = e.message + info
+    e.extra_info.message = e.extra_info.message + info
+    e.args = (e.args[0], e.args[1] + info)
```

Comment:
Second, let's never mess with `args`.

```
>>> class MyError(Exception):
...     def __init__(self, *args):
...             pass
... 
>>> MyError(1, 2)
MyError(1, 2)
>>> e = MyError(1, 2)
>>> e.
e.args             e.with_traceback(  
>>> e.args
(1, 2)
```

`args` exception attribute is essentially a copy of "what data was used to raise exception with".

Reply:
done

Reply:
Done, deleted